battle-service/src/main/java/by/ibondarau/tictactoe/battleservice/service/impl/BattleServiceImpl.java

```diff
@@ -0,0 +1,144 @@
+
+    @Override
+    public Battle createBattle(UUID playerId, int size, @NotNull FirstMoveRule firstMoveRule) {
+        Battle battle = new Battle()
+                .setFirstPlayerId(playerId)
+                .setSize(size)
+                .setFirstMoveRule(firstMoveRule)
+                .setStatus(BattleStatus.CREATED);
+        battleDao.save(battle);
+        return battle;
+    }
+
+    @Override
+    public Battle getBattle(UUID battleId) {
+        return battleDao
+                .findBattleByBattleId(battleId)
+                .orElseThrow(() -> new NotFoundException("Battle with id " + battleId + " not found"));
+    }
+
+    @Override
+    public Battle joinBattle(UUID battleId, UUID secondPlayerId) {
+        Battle battle = battleDao.findBattleByBattleId(battleId).orElseThrow(() -> new NotFoundException("Battle with id " + battleId + " not found"));
+
+        if (!BattleStatus.CREATED.equals(battle.getStatus()) || battle.getSecondPlayerId() != null) {
+            throw new BusinessException("Unable to join: invalid battle status");
+        }
+
+        battle.setSecondPlayerId(secondPlayerId);
+        battle.setStatus(BattleStatus.STARTED);
+        battle.setStarted(Timestamp.from(ZonedDateTime.now().toInstant()));
+        setFirstMovingPlayer(battle);
+        battle = battleDao.save(battle);
+        return battle;
+    }
+
+    private void setFirstMovingPlayer(Battle battle) {
+        battle.setFirstMovingPlayerId(switch (battle.getFirstMoveRule()) {
+            case FIRST_GOES_FIRST -> battle.getFirstPlayerId();
+            case SECOND_GOES_FIRST -> battle.getSecondPlayerId();
+            case RANDOM -> randomUtils.getRandomOf(battle.getFirstPlayerId(), battle.getSecondPlayerId());
+        });
+    }
+
+    @Override
+    public Battle makeMove(UUID battleId, Move move) {
+        Battle battle = battleDao.findBattleByBattleId(battleId).orElseThrow(() -> new NotFoundException("Battle with id " + battleId + " not found"));
+        addMove(battle, move);
+        updateGameState(battle);
+        battle = battleDao.save(battle);
+        return battle;
+    }
+
+    private void addMove(Battle battle, Move move) {
+        UUID nextMovePlayerId = battleUtils.getNextMovePlayerId(battle);
+        if (nextMovePlayerId == null || !nextMovePlayerId.equals(move.getPlayerId())) {
+            throw new BusinessException("Bad move: invalid player id");
+        }
+        if (move.getFirstCoordinate() >= battle.getSize() || move.getSecondCoordinate() >= battle.getSize()) {
+            throw new BusinessException("Bad move: coordinates are out of field bounds");
+        }
+
+        if (battle.getMoves().stream().anyMatch(battleMove ->
+                battleMove.getFirstCoordinate().equals(move.getFirstCoordinate())
+                        && battleMove.getSecondCoordinate().equals(move.getSecondCoordinate())
+        )) {
+            throw new BusinessException("Bad move: cell is not empty");
```

Comment:
this BusinessException seems to be a good "parent" for MoveException

Reply:
Done.